include/odp/api/spec/packet.h

```diff
@@ -1140,6 +1179,82 @@ int odp_packet_move_data(odp_packet_t pkt, uint32_t dst_offset,
  */
 
 /**
+ * Packet parse parameters
+ */
+typedef struct odp_packet_parse_param_t {
+	/** Protocol header at parse starting point. Valid values for this
+	 *  field are: ODP_PROTO_ETH, ODP_PROTO_IPV4, ODP_PROTO_IPV6. */
+	odp_proto_t proto;
+
+	/** Continue parsing until this layer. Must be the same or higher
+	 *  layer than the layer of 'proto'. */
+	odp_proto_layer_t layer;
+
+	/** Flags to control payload data checks up to the selected parse
+	 *  layer. Checksum checking status can be queried for each packet with
+	 *  odp_packet_l3_chksum_status() and odp_packet_l4_chksum_status().
+	 */
+	union {
+		struct {
+			/** Check IPv4 header checksum */
+			uint32_t ipv4_chksum   : 1;
+
+			/** Check UDP checksum */
+			uint32_t udp_chksum    : 1;
+
+			/** Check TCP checksum */
+			uint32_t tcp_chksum    : 1;
+
+		} check;
+
+		/** All check bits. This can be used to set/clear all flags. */
+		uint32_t all_check;
+	};
+
+} odp_packet_parse_param_t;
+
+/**
+ * Parse packet
+ *
+ * Parse protocol headers in packet data. Parsing starts at 'offset', which
+ * is the first header byte of protocol 'param.proto'. Parameter 'param.layer'
+ * defines the last layer application is interested about.
+ * Use ODP_PROTO_LAYER_ALL for all layers. The operation sets or resets packet
+ * metadata for all layers from the layer of 'param.proto' to the application
+ * defined last layer. Metadata of other layers have undefined values.
```

Comment:
What about: "Metadata of other layers remains unmodified."?

Reply:
Application calls parse because metadata is not valid to begin with. There would not be much benefit for application to require that implementation saves old metadata values before the operation and then restores those after if e.g. an internal error occurred. Implementation performance would be hurt by making a copy of the old metadata.

Since it's just (L2/L3/L4) metadata (not data) that may have changed, application can continue processing the packet data also after a failed parse.

Reply:
Processing of UDP/TCP (in OFP and BSD like) depends on IPv4/IPv6: there are different input functions for UDP/TCP processing for IPv4 and IPv6 (https://github.com/OpenFastPath/ofp/blob/master/src/ofp_in_proto.c, https://github.com/OpenFastPath/ofp/blob/master/src/ofp_in6_proto.c)